### PR TITLE
Rmlui integration part 1: Core

### PR DESCRIPTION
**Pull request by ChrisFloofyKitsune**
Integration for RmlUi library (minus the Lua-related stuff)
Changes carefully copied from https://github.com/loveridge/spring/tree/rmlui-integration-fork

The credit for this mostly goes to these two:
- [p2004a](https://github.com/p2004a)
- @loveridge 

I'm just the fox that was annoyed at the in-game info UIs being so cluttered and wound up bringing this project that much closer to the finish line.

------

The RmlUi integration here is mostly along these lines:
- Add RmlUi and lunasvg to the build
- start/shutdown RmlUi with game start/shutdown
- Implement the needed System, File, and Render Interfaces from RmlUi to have it communicate with the game engine and the VFS.
- feed appropriate events to RmlUi
  - make sure that when RmlUi sets the cursor, it actually shows up
- render the RmlUi contexts after rendering the Lua Ui

-----

Due to the removal of Lua related changes, there are no working RmlUi gui widgets to look at. :(

(Not that there was much else than the barest minimum needed to test that RmlUi was working properly.)

-----

The <lua-texture> element that's still included from the working branch I left in because it doesn't interact with the Lua runtime at all. It just hooks into the same mechanism that the Lua side of things uses to load/reuse textures.

I can remove or rename it. 

It also may be helpful as an example for other custom elements in the future.

### CODE REVIEW

**Comment by lhog** [comment]
https://github.com/beyond-all-reason/spring/blob/1a0fb03c0366820ae299aa630569d582f8b79e8e/rts/Rml/Backends/RmlUi_Renderer_GL3_Spring.cpp should be redone, starting from the name of the file (the project is called Recoil), but most importantly all GL primitives like VAO/VBO, Fixed state, shaders and alike have engine-side encapsulation and convenience methods.
Do you feel you can do it yourself or should I take this task post merging?

File: rts/Game/UI/GuiHandler.cpp [hunk 1]
```diff
@@ -168,6 +169,8 @@ bool CGuiHandler::EnableLuaUI(bool enableCommand)
 		}
 	}
 
+	// TODO: Move to own console command?
+	RmlGui::Reload();
```
Comment (lhog) [inline rts/Game/UI/GuiHandler.cpp:173]: What does `RmlGui::Reload()` effectively do? I've read the function's code, but I'm interested in user-facing effects.

Comment (ChrisFloofyKitsune) [reply rts/Game/UI/GuiHandler.cpp:173]: By reloading the RmlUi library, cached data is cleared.

Stylesheets and images most notably.

Of course something then needs to notify the code to reload the Rml based guis (typically a Widget from the LuaUI environment)...

Seems that the RmlGui::Reload() call needs to stay where it is then.

Reloading LuaUI or Rml separately from the other looks like it's just gonna break things

**Comment by ChrisFloofyKitsune** [comment]
> https://github.com/beyond-all-reason/spring/blob/1a0fb03c0366820ae299aa630569d582f8b79e8e/rts/Rml/Backends/RmlUi_Renderer_GL3_Spring.cpp should be redone, starting from the name of the file (the project is called Recoil), but most importantly all GL primitives like VAO/VBO, Fixed state, shaders and alike have engine-side encapsulation and convenience methods. Do you feel you can do it yourself or should I take this task post merging?

I'd like to take a crack at it, it'll be helpful experience for getting the Lua integration properly working.

**Comment by ChrisFloofyKitsune** [comment]
> https://github.com/beyond-all-reason/spring/blob/1a0fb03c0366820ae299aa630569d582f8b79e8e/rts/Rml/Backends/RmlUi_Renderer_GL3_Spring.cpp should be redone, starting from the name of the file (the project is called Recoil), but most importantly all GL primitives like VAO/VBO, Fixed state, shaders and alike have engine-side encapsulation and convenience methods. Do you feel you can do it yourself or should I take this task post merging?

I've pushed up the rework of the Render Interface now

File: rts/Rml/Backends/RmlUi_Backend.cpp [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+#include "Rml/RmlInputReceiver.h"
+#include "RmlUi_Backend.h"
+#include "RmlUi_Renderer_GL3_Recoil.h"
+#include "RmlUi_SystemInterface.h"
+#include "RmlUi_VFSFileInterface.h"
+#include "System/Input/InputHandler.h"
+#include "System/Log/ILog.h"
+
+using CtxMutex = std::recursive_mutex;
+using CtxLockGuard = std::lock_guard<CtxMutex>;
```
Comment (lhog) [inline rts/Rml/Backends/RmlUi_Backend.cpp:50]: Why do we need a mutex here?

File: rts/Rml/Backends/RmlUi_Backend.cpp [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+	// make atomic_bool?
+	bool initialized = false;
+	bool debuggerAttached = false;
+	int winX = 1;
+	int winY = 1;
+
+	lua_State* ls = nullptr;
+
+	CtxMutex contextMutex;
+	Rml::UniquePtr<PassThroughPlugin> plugin;
```
Comment (lhog) [inline rts/Rml/Backends/RmlUi_Backend.cpp:114]: Replace with `std::` counterparts

**Comment by lhog** [comment]
Moved to https://github.com/beyond-all-reason/spring/pull/1304

File: rts/Rml/Backends/RmlUi_Backend.cpp (hunk 2 above)
Comment (ChrisFloofyKitsune) [reply rts/Rml/Backends/RmlUi_Backend.cpp:50]: mostly to make sure that a lua script removing a context at an arbitrary point does not cause a crash

**Pull request merged**